### PR TITLE
Projection version advancing fix

### LINKED ISSUE #583: Fix Projection version increments
Right now, all the `Entity` versions are incremented by injecting the version from `EventContext`.
But when talking about `Projection`s, the event versions are totally irrelevant.

That is why we need to increment the version of the `Projection`s, instead of injecting them.

### PR DESCRIPTION
**Pull request by dmdashenkov**
Fixes #583.

From now there are two strategies of version advancing: `FROM_EVENT` and `AUTO_INCREMENT`.

The `FROM_EVENT` strategy is default, and the `AUTO_INCREMENT` is used for `Projection`s and `ProcessManager`s.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: server/src/main/java/io/spine/server/entity/AbstractVersionableEntity.java [hunk 1]
```diff
@@ -173,6 +174,7 @@ private void updateVersion(Version newVersion) {
      *
      * @param newState a new state to set
      */
+    @VisibleForTesting
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/AbstractVersionableEntity.java]: I think this annotation can be removed. Let's discuss if needed, but it shouldn't be here.

File: server/src/main/java/io/spine/server/entity/Transaction.java [hunk 2]
```diff
@@ -445,6 +439,61 @@ public void setDeleted(boolean deleted) {
...
+     * <p>The default value is {@link VersionAdvancementStrategy#INJECT INJECT}.
+     */
+    protected VersionAdvancementStrategy versioningStrategy() {
+        return INJECT;
+    }
+
+    /**
+     * The strategy of advancing the version of a {@code Transaction} upon the given {@link Phase}.
+     */
+    protected enum VersionAdvancementStrategy {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/Transaction.java]: `EntityVersioning`

And please pull it to a package level.

File: server/src/main/java/io/spine/server/entity/Transaction.java [hunk 3]
```diff
@@ -445,6 +439,61 @@ public void setDeleted(boolean deleted) {
...
+     * <p>The value should be constant along all the instances of a certain (runtime) type.
+     *
+     * <p>The default value is {@link VersionAdvancementStrategy#INJECT INJECT}.
+     */
+    protected VersionAdvancementStrategy versioningStrategy() {
+        return INJECT;
+    }
+
+    /**
+     * The strategy of advancing the version of a {@code Transaction} upon the given {@link Phase}.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/Transaction.java]: `The strategy of versioning of entities during a phase of a certain transaction.` (just add proper Javadoc links etc).

File: server/src/main/java/io/spine/server/entity/Transaction.java [hunk 4]
```diff
@@ -445,6 +439,61 @@ public void setDeleted(boolean deleted) {
     }
 
     /**
+     * Retrieves the {@link VersionAdvancementStrategy} of current {@code Transaction}.
+     *
+     * <p>The value should be constant along all the instances of a certain (runtime) type.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/Transaction.java]: `among all`

File: server/src/main/java/io/spine/server/entity/Transaction.java [hunk 5]
```diff
@@ -445,6 +439,61 @@ public void setDeleted(boolean deleted) {
...
+     *
+     * <p>The default value is {@link VersionAdvancementStrategy#INJECT INJECT}.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/Transaction.java]: Please explain why, not just duplicate the implementation.

File: server/src/main/java/io/spine/server/entity/Transaction.java [hunk 6]
```diff
@@ -445,6 +439,61 @@ public void setDeleted(boolean deleted) {
...
+     */
+    protected enum VersionAdvancementStrategy {
+
+        /**
+         * Retrieves the version from the {@linkplain EventContext event}, which caused
+         * the transaction phase.
+         *
+         * @see EventContext#getVersion()
+         */
+        INJECT {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/Transaction.java]: This is `FROM_EVENT`.

File: server/src/main/java/io/spine/server/entity/Transaction.java [hunk 7]
```diff
@@ -445,6 +439,61 @@ public void setDeleted(boolean deleted) {
...
+                return fromEvent;
+            }
+        },
+
+        /**
+         * Retrieves the current version of the {@code Transaction} incremented by one.
+         *
+         * @see Transaction#getVersion()
+         */
+        INCREMENT {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/Transaction.java]: This is `AUTO_INCREMENT`. This is a well-known term for this behaviour.

File: server/src/main/java/io/spine/server/entity/Transaction.java [hunk 8]
```diff
@@ -445,6 +439,61 @@ public void setDeleted(boolean deleted) {
...
+        INJECT {
+            @Override
+            Version advanceUpon(Phase<?, ?, ?, ?> phase) {
+                final Version fromEvent = phase.getContext().getVersion();
+                return fromEvent;
+            }
+        },
+
+        /**
+         * Retrieves the current version of the {@code Transaction} incremented by one.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/Transaction.java]: The description must be better. This is an enum member. It does not retrieve anything. Same applies to the previous enumeration member.

File: server/src/main/java/io/spine/server/entity/Transaction.java [hunk 9]
```diff
@@ -445,6 +439,61 @@ public void setDeleted(boolean deleted) {
...
+            @Override
+            Version advanceUpon(Phase<?, ?, ?, ?> phase) {
+                final Version current = phase.getUnderlyingTransaction().getVersion();
+                final Version newVersion = Versions.increment(current);
+                return newVersion;
+            }
+        };
+
+        /**
+         * Creates an advanced version upon the given {@link Phase}.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/Transaction.java]: I don't understand. Let's discuss.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: server/src/main/java/io/spine/server/entity/Transaction.java [hunk 10]
```diff
@@ -444,6 +439,18 @@ public void setDeleted(boolean deleted) {
     }
 
     /**
+     * Retrieves the {@link EntityVersioning} of current {@code Transaction}.
+     *
+     * <p>The value should be constant among all the instances of a certain (runtime) type.
+     *
+     * <p>The default value is the Event Sourcing-fashioned
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/Transaction.java]: `By default the version is taken from the {@linkplain ... latest event applied}`.

File: server/src/main/java/io/spine/server/entity/EntityVersioning.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        Version nextVersion(Transaction.Phase<?, ?, ?, ?> phase) {
+            final Version current = phase.getUnderlyingTransaction().getVersion();
+            final Version newVersion = Versions.increment(current);
+            return newVersion;
+        }
+    };
+
+    /**
+     * Creates the {@link Entity} version which is set after the given {@link Transaction.Phase} is
+     * complete successfully.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/EntityVersioning.java]: `completed`.

File: server/src/main/java/io/spine/server/entity/EntityVersioning.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    FROM_EVENT {
+        @Override
+        Version nextVersion(Transaction.Phase<?, ?, ?, ?> phase) {
+            final Version fromEvent = phase.getContext().getVersion();
+            return fromEvent;
+        }
+    },
+
+    /**
+     * This strategy is applied to the {@link Entity} types which are not the pure Event
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/EntityVersioning.java]: This should be rephrased.

File: server/src/main/java/io/spine/server/entity/EntityVersioning.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+     * Sourcing entities, such as {@link io.spine.server.projection.Projection Projection}s.
+     *
+     * <p>A {@code Projection} represents an arbitrary cast of data in a specific moment in
+     * time. The events applied to a {@code Projection} are random, i.e. are produced by
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/EntityVersioning.java]: They are not random. So you may as well just delete this portion.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: server/src/test/java/io/spine/server/projection/ProjectionTransactionShould.java [hunk 14]
```diff
@@ -140,6 +146,30 @@ protected void breakEntityValidation(
               .setShouldThrow(toThrow);
     }
 
+    @Ignore // The behavior is changed. See increment_version_on_event for the right behavior test
```
Comment (armiol) [inline server/src/test/java/io/spine/server/projection/ProjectionTransactionShould.java]: It should be:
```
/**
 * This test is ignored as the expected behavior has been changed for the projection transactions.
 * 
 * <p>{@link #increment_version_on_event() Another test method} is created to test the new behavior. 
 * Please refer to it for more details.
 */
```

And please mention that this test hierarchy should be eliminated in the PM versioning issue you create(d).

File: server/src/test/java/io/spine/server/projection/ProjectionTransactionShould.java [hunk 15]
```diff
@@ -140,6 +146,30 @@ protected void breakEntityValidation(
...
+    @Test
+    @Override
+    public void advance_version_from_event() { }
```
Comment (armiol) [inline server/src/test/java/io/spine/server/projection/ProjectionTransactionShould.java]: Please delete an extra space.

**Pull request merged**